### DISCUSSION TITLE
Real Time Capability

### DISCUSSION THREAD

**Discussion by LorenzLaubach** [dstl/Stone-Soup#446] [Q&A]
Hello, i am using the stonesoup framework for my masterthesis project, where i have to fuse sensor data.
I am using a sensor to global Fusion Strategy. So i have a global Object list and each Sensor gives me a new Object List each time stamp. I want to use the framework to associate the new incoming Objects with maintained global Objects and then fuse them together with a kalman filter, but i am struggling since my code cannot keep up with the incoming data stream. 
So the Question is if Stonesoup in General is capable to use it for real time problems.
I am currently using the GNN Method for Association and the simple Kalman Filter as the Updater.
I have noticed that at the Hypothesise step the "predict_measurement" step needs quite long to be calculated. What exactly is happening there? I thought it would be possible to create Hypothesis based alone of the distances of the predicted Track (my global Object) and Measurement (Incoming Objects by the Sensors). 

I would be very thankfull for every kind of help!

Best regards Lorenz

**Comment by sdhiscocks** [comment]
> So the Question is if Stonesoup in General is capable to use it for real time problems.

So Stone Soup is designed more for assessing algorithms and having flexible framework, so may not run real time for your problem. However, often it can run fast enough for real time; again depending on filter/models/metrics being used.

We are always interested in ways of speeding up Stone Soup.

> I am currently using the GNN Method for Association and the simple Kalman Filter as the Updater.

Data association is often a bottle neck. Are you using the [`GlobalNearestNeighbour`](https://stonesoup.readthedocs.io/en/v0.1b5/stonesoup.dataassociator.html#stonesoup.dataassociator.neighbour.GlobalNearestNeighbour) class data associator? This is horribly slow, because it actually calculates all combinations. [`GNNWith2DAssignment`](https://stonesoup.readthedocs.io/en/v0.1b5/stonesoup.dataassociator.html#stonesoup.dataassociator.neighbour.GNNWith2DAssignment) is a straight swap replacement, and much much faster as it uses an 2D assignment algorithm. 

> I have noticed that at the Hypothesise step the "predict_measurement" step needs quite long to be calculated. What exactly is happening there? 

The predicted measurement step is a conversion of the the tracks' state space, to the measurements' state space. This is required in order to compare the track to measurement. The speed of this will be dependent on your measurement model.

>  I thought it would be possible to create Hypothesis based alone of the distances of the predicted Track (my global Object) and Measurement (Incoming Objects by the Sensors).

So one option would be to create a custom [`Gater`](https://stonesoup.readthedocs.io/en/v0.1b5/stonesoup.gater.html#stonesoup.gater.base.Gater) class, which compare elements of the track state space to measurement state space in a way to avoid calling `Update.predict_measurement`.

As another thought, we hope to merge #145 soon as well, which can really help (and significantly) in some scenarios.
